Add support for h2d file format with first data file

ihhub opened the pull request:
- fix farm rendering in Knight castle (#2059)
- add missing farm sprite made by our artist (close #1080 )

![image](https://user-images.githubusercontent.com/19829520/125189668-ccc87780-e26b-11eb-8fbf-1e01206c19a6.png)

ihhub commented:
Hi @oleg-derevenetz and @idshibanov , this is the first step for storing any data files for the upcoming add-on so could you please light out your opinion about the way how we can store files?

oleg-derevenetz commented:
> Hi @oleg-derevenetz and @idshibanov , this is the first step for storing any data files for the upcoming add-on so could you please light out your opinion about the way how we can store files?

Hi @ihhub Is there a reason to store them as a separate files? Why not store them in binary itself as a some sort of resources (for example, as trivial `const char *` or `std::array<char>` arrays)?

ihhub commented:
> > Hi @oleg-derevenetz and @idshibanov , this is the first step for storing any data files for the upcoming add-on so could you please light out your opinion about the way how we can store files?
> 
> Hi @ihhub Is there a reason to store them as a separate files? Why not store them in binary itself as a some sort of resources (for example, as trivial `const char *` or `std::array<char>` arrays)?

In the future we're going to add much more resources including new monsters or sounds for example. We're talking about megabytes of data. I don't think that keeping them in binary would be the best solution.

oleg-derevenetz commented on src/fheroes2/h2d/h2d_file.cpp:
```diff
@@ -0,0 +1,173 @@
+        const int32_t height = static_cast<int32_t>( stream.getLE32() );
+        const int32_t x = static_cast<int32_t>( stream.getLE32() );
+        const int32_t y = static_cast<int32_t>( stream.getLE32() );
+        if ( static_cast<size_t>( width * height * 2 + 4 + 4 + 4 + 4 ) != data.size() ) {
```
I suppose it would be good to check `x` and `y` as well - they should be non-negative and less than `width` and `height` respectively? Also beware of signed overflow when operating with values from an external source - at least the stage of checking the values in a good way should take into account this possibility (yes, I know that in reality, the consequences will be minimal, but it's still an UB).

ihhub replied:
x and y values can be negative. They don't relate to the current image but to the image on which we are drawing the on. For example, the farm image has coordinates in relation to town's background image. However, for width and height we have to check values.

oleg-derevenetz replied:
Ah, OK.

oleg-derevenetz commented on src/fheroes2/h2d/h2d.cpp:
```diff
@@ -0,0 +1,63 @@
+
+        isInitialized = true;
+
+        ListFiles files = Settings::FindFiles( "data", ".h2d", false );
```
The choice of the "data" directory is generally logical, but the fact that the original game uses the same directory is a bit confusing I think. This directory may be accidentally removed or replaced by user. May be rename it somehow? Or may be move it to "files", as it done with "fonts" and "lang"?

ihhub replied:
You have a valid point. Let's put then to files. Also I'm thinking since we have to provide translation files we can put them into h2d as well. I will create a separate todo issue with detailed description for translation. What do you think?

oleg-derevenetz replied:
> Also I'm thinking since we have to provide translation files we can put them into h2d as well.

I'm not very familiar with gettext, so at the moment I can't say if there are any advantages to it :)

oleg-derevenetz reviewed (commented):
Hi @ihhub LGTM in general, left just two notes.

ihhub commented:
Hi @oleg-derevenetz , looks like we have to disable `cppcoreguidelines-pro-bounds-pointer-arithmetic`, `modernize-use-auto` and `cppcoreguidelines-pro-type-reinterpret-cast` clang-tidy checks.

oleg-derevenetz commented:
> looks like we have to disable `cppcoreguidelines-pro-bounds-pointer-arithmetic`, `modernize-use-auto` and `cppcoreguidelines-pro-type-reinterpret-cast` clang-tidy checks.

Yes, I agree regarding the first two checks, but regarding the `cppcoreguidelines-pro-type-reinterpret-cast`... This cast is really dangerous (although it may be used when casting to `char*`) so may be leave it and just check twice on warning and if everything is OK just ignore it?

ihhub commented:
> @ihhub
> 
> > looks like we have to disable `cppcoreguidelines-pro-bounds-pointer-arithmetic`, `modernize-use-auto` and `cppcoreguidelines-pro-type-reinterpret-cast` clang-tidy checks.
> 
> Yes, I agree regarding the first two checks, but regarding the `cppcoreguidelines-pro-type-reinterpret-cast`... This cast is really dangerous (although it may be used when casting to `char*`) so may be leave it and just check twice on warning and if everything is OK just ignore it?

Sure. Let's do this way.

ihhub commented:
Hi @oleg-derevenetz , I addressed your comment. Could you please take a look?

Right now it's not obvious why we need h2d file but in the future with add-on development (which started not long time ago) we have to ship all new assets altogether. To reduce user errors such missing files during copying.

ihhub commented:
Hi @Branikolog , could you please also test these changes? Just install a fresh version of the game and check a town with the farm in Knight castle.

Branikolog commented:
Hi, @ihhub 
Everything seems ok. Farm is fully drawn. Castle is being built correctly over it.
Only a few remarks I can add here:
New part of a farm is not clickable. We've already have an issue with farm, as we should click only small area inside it to initiate info window to appear (As well as may other castle objects). So now this too tiny area becomes even more obvious.
Also, I think in future, we can change farm icon in castle building window, so it would introduce full field sprite, while the original icon shows only part:
![image](https://user-images.githubusercontent.com/55348946/129473401-ae072aac-5a99-4a17-9cd5-828bed210e4c.png)
